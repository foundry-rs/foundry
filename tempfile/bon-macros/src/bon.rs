use crate::builder;
use crate::normalization::{ExpandCfg, Expansion};
use crate::util::prelude::*;
use darling::ast::NestedMeta;
use darling::FromMeta;

pub(crate) fn generate(params: TokenStream, item: TokenStream) -> TokenStream {
    crate::error::handle_errors(item.clone(), || try_generate(params, item))
        .unwrap_or_else(std::convert::identity)
}

pub(crate) fn try_generate(params: TokenStream, item: TokenStream) -> Result<TokenStream> {
    let item: syn::Item = syn::parse2(item)?;

    let ctx = ExpandCfg {
        current_macro: format_ident!("bon"),
        config: params,
        item,
    };

    let input = match ctx.expand_cfg()? {
        Expansion::Expanded(input) => input,
        Expansion::Recurse(output) => return Ok(output),
    };

    let params = NestedMeta::parse_meta_list(input.config)?;
    let params = FromMeta::from_list(&params)?;

    match input.item {
        syn::Item::Impl(item_impl) => builder::item_impl::generate(params, item_impl),
        _ => bail!(
            &input.item,
            "`#[bon]` attribute is expected to be placed on an `impl` block \
             but it was placed on other syntax instead"
        ),
    }
}
