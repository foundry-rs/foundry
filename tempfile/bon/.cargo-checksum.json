{"files":{"Cargo.toml":"9ab0d2f0f5c7dc8bd712d585ef5709736b508827406381e82336c6c954f505e3","README.md":"fe8f6820217faf300ad66c626d0bc2f858cbda1a880cfa27f77d5f49d3013d3f","src/__/better_errors.rs":"d802d5b1da46174e5409f2677c0f0e559b8902eca5cba0e12f1eb10054bfb1e4","src/__/cfg_eval.rs":"a861a125b1eb0fe8927051e52b31f5f24ea93589dcddb80888fec015a42d0be9","src/__/ide.rs":"0f6bccd0cead088b5ed570b258d7afe53b3c5d34166a91a065d3bb52ec0f5360","src/__/mod.rs":"663588477d8485763c2c21ececb928269de5a9d03c4bb8670413cd7b77c779d2","src/builder_state.rs":"6b3bc0efcaf3a188648d743fa59c809eb18e983ee2fac60f71369070658b44f0","src/collections.rs":"d2e2600fe79218ea27b4c6b08d8e79f60b07e208dac06567f638d1509f314117","src/lib.rs":"c5f0658e223bf99dfa4f0bf5798a7bb79c6a0874cccaba417c6824d143dc189d","tests/integration/builder/attr_bon.rs":"44644967967d8b91e8f1c1bc3c5491b39d5d8bef9b41d0c4747e5e604d5b5207","tests/integration/builder/attr_crate.rs":"0978828b63ccc23383abbf13ba7113a5f29e806c5e1a7a3e0779225405439ec1","tests/integration/builder/attr_default.rs":"ea95f8ae91ffed3484af863c52cde11c46447f6acb87396244b313794af8940d","tests/integration/builder/attr_derive.rs":"f93ac31a0984761cab11769765eaf04b57321c77bd086f31bbfadf35088edf4b","tests/integration/builder/attr_field.rs":"b09f360697c17d5a711c85f985b0ce746d385f1af0347e22da68f3e8b607f8dc","tests/integration/builder/attr_getter.rs":"517ec0b19c16855ae041042c385a55b24a64da5e5dd21800a105505b10417bbf","tests/integration/builder/attr_into.rs":"29f39269d5e0f6e705a2b703cf625bd3ca7280762445fd00360717171120d4eb","tests/integration/builder/attr_on.rs":"6894da21a448f6930fd99bfd172a16690a6927602e0e1d674d52c07109c88cf0","tests/integration/builder/attr_overwritable.rs":"fb472587caaffc6964aa0fa559e7b820086e4e32e865fd10bffdbef6a59aab6b","tests/integration/builder/attr_required.rs":"18ae48334d77996fb760531449a73328a4b4241e0dfa60aecfbdc31a6f37e3ef","tests/integration/builder/attr_setters.rs":"a83d1e0e291f3afbe877186ac7d73de70c474bb69730d3aa090ab56373c34053","tests/integration/builder/attr_skip.rs":"16e05897185756d21a68fbc55b59eb2231cf4973fb2f1bd95b2f8e485ad9f656","tests/integration/builder/attr_top_level_start_fn.rs":"2e408c3d5b4d2e101a613b263c03846d82eb3b63b8ae1e6c8899bbae816042cd","tests/integration/builder/attr_with/from_iter.rs":"425d835f1768b70ad3468f14aa2a29e6f1d03015472b2758889e355b24d6fff3","tests/integration/builder/attr_with/mod.rs":"d0267099b3fc1090d55e5af33127712e7e26084f6ec2cbdf0da4139323869f54","tests/integration/builder/attr_with/multi_arg.rs":"97b7555b634e1120f031281014122af8daebc02c71fbf23ed7fa084ef0fcb5dc","tests/integration/builder/attr_with/overwritable.rs":"716f311b425574b5184a681cfc97c6f7cf934d8bb74f1a03fb8bc6def3f2f4ae","tests/integration/builder/attr_with/single_arg.rs":"eaa4ca521b0e4d12697e34fa0f9a8fb0b0b18f6b7023174a85a8c859827f00d4","tests/integration/builder/attr_with/some.rs":"a797e5a8e59744f85fafde1f293ad9ae4e81c88f8a4457ef4ac1aa36e8c1ea65","tests/integration/builder/cfgs.rs":"751ce8f087ad3104e3cfddbb290a4fb93394349539af399a9feced3bbf741b90","tests/integration/builder/generics.rs":"254e03f60f0224879575f0d610973fbb125b390fc18337c36bfbe8f9e7d5fac8","tests/integration/builder/init_order.rs":"01609bb2f4ca0838ada26b360bc9508d4e059328744ae4ec533e2d17b04e5406","tests/integration/builder/lints.rs":"f293bc93b7057f0e60556f3b9ce5b6f33386b9f14e4e79201245b005e177e5fe","tests/integration/builder/many_params.rs":"59d35ce48301d67c0e7a51f6d152b434704f417eed835cbb8b79563d99169b4f","tests/integration/builder/mod.rs":"421f319e1f57881c403e74fa34be9808ccffb356fce8926374d80b4c5af589d6","tests/integration/builder/name_conflicts/builder_state.rs":"8c6d8006973aef90db99f90c1f9cfbf0a37d5dc1d7bd4eeff42083334280b556","tests/integration/builder/name_conflicts/generics.rs":"fb0fe9292a9e74918992f727978ed56b8bc539c4132cdbe11d4236b23ca68e91","tests/integration/builder/name_conflicts/member_and_type_named_the_same.rs":"781fdd528ea201f899a493a7d7625c46a69979bf6974be782380c2b03e5039af","tests/integration/builder/name_conflicts/member_names_state.rs":"2837b833be72a2109aed2dcf0214e95127ae1ffd305c645f6c9ec664f06e86d3","tests/integration/builder/name_conflicts/mod.rs":"ebf10d20bd977383839e5d0792a334fcab349fe2c4236c2558082bc6175d3b79","tests/integration/builder/positional_members.rs":"f35d5b5532b81816923fcd66057218ef5454a4600863ba2c9a523e44fa84eeb1","tests/integration/builder/raw_idents.rs":"0eed39a82dd4e9dd15d06438562f937e77a9394652cfd1b2a3b6a7dfc78bd27d","tests/integration/builder/smoke.rs":"a6ec81faa4b3a9292e45ae8d93e79749d4cb690b8ba1a0ec67961d31d701e7d2","tests/integration/main.rs":"cd7f9feb0825d9451d5cf8d104f85a2aa2b80aced23e8c8d8557b762ec0374b5","tests/integration/ui/compile_fail/attr_bon.rs":"3b601a5d176ee41d7d6d52855f3d5548a63f13bef5dca435bf7aa6dba12a878a","tests/integration/ui/compile_fail/attr_bon.stderr":"38faf97b17aa923d0c7143a381eff4293483e42e0fbdb3f3eb200716d38468ef","tests/integration/ui/compile_fail/attr_builder.rs":"ec6d77915d90b0e22a3ef7a0e99953dcf70a543f2a571fa266e0a27cf571f3e8","tests/integration/ui/compile_fail/attr_builder.stderr":"ea7b100cb15f6e0b0e2a49d32dd64619bc84c02cce8957cc1e75ec5d16dbd868","tests/integration/ui/compile_fail/attr_crate.rs":"accca190fdd8cfca3da478ed097afbda6a20e056336127d812386c6afaf65b6f","tests/integration/ui/compile_fail/attr_crate.stderr":"d1fb6c9c2dedb2120c8a4c39da97c1c524d5019ccdda8479d98ef31a950eb3e3","tests/integration/ui/compile_fail/attr_derive.rs":"644328fc8007c4594e70f264d35bb1c9ae6ebd15e3c4511fcf136ee2e461a723","tests/integration/ui/compile_fail/attr_derive.stderr":"44a677981cd300c5bc3ad76975e59094f9dba92e504d2b68ccfe739e9b55d638","tests/integration/ui/compile_fail/attr_field.rs":"e5b6d895a204bee2c0a749c49c2ef817c46f0ae827c85129f61b7f56b24172c4","tests/integration/ui/compile_fail/attr_field.stderr":"1007f8d50af2866c34e1577b7773c40385183002c221ff1acf344f2fc7e32698","tests/integration/ui/compile_fail/attr_getter.rs":"9790604b53b741debcb3d4aef45a2f7edf515803c4f1c9929bb03ee655cb6dcc","tests/integration/ui/compile_fail/attr_getter.stderr":"e5a28a51f8d2f9fcc423339f06e6b1c53f8b211b31146f825309540c38b63cff","tests/integration/ui/compile_fail/attr_into.rs":"9c2ec0ff0d49161f494f9029471bedf4e441a12fcf424a83ef4413c6503c7c75","tests/integration/ui/compile_fail/attr_into.stderr":"a02d5962fabe0d02bc7cd762cbe67e69cbcbeed0c5fad9bbbfe6202ade7b1cba","tests/integration/ui/compile_fail/attr_on.rs":"b20df7bdc8857c4040a4d0675b1dd92fbbeb93f179719c024a253cc9d38109cf","tests/integration/ui/compile_fail/attr_on.stderr":"2cdd5fa0d3a75a8464833dd2aac790e5a3d31894aa46d19df7c6998c5cf18e58","tests/integration/ui/compile_fail/attr_required.rs":"f7bd10614e4e4b5e2480fb2d41f6efc8bb917986714d131e9506967e40578433","tests/integration/ui/compile_fail/attr_required.stderr":"0597bee0f4d1beb3eae512d78c73c66230fcd49e75e95d5e1035471853c1f485","tests/integration/ui/compile_fail/attr_setters.rs":"1ce210867b4492f57106263807bc08d038480b26785369a7cb579f06b4dce3b7","tests/integration/ui/compile_fail/attr_setters.stderr":"4062c87bd360a4c07b7def7049e5c6e69d720467d77dbfe32b777529fc2d876a","tests/integration/ui/compile_fail/attr_skip.rs":"b063693f7fbf069a7299e2b70b1c52808f148f51e133fcf1545fd61cb277c449","tests/integration/ui/compile_fail/attr_skip.stderr":"1cfdc9b870f5c8c8e8d10a707aa86e8370ce33e1e10b9954d5f79dbddb817318","tests/integration/ui/compile_fail/attr_top_level_start_finish_fn.rs":"bec69ca91d02043360cf9591adac1e8d377a877c74b44b30d24bc86ac3c8b8c5","tests/integration/ui/compile_fail/attr_top_level_start_finish_fn.stderr":"810bf95812605d5c927f670e6a9f627eef1b31e4c145fd7540106d83bedaf92f","tests/integration/ui/compile_fail/attr_with.rs":"edaeef4a173ac2ace271feeccc1f2f1a129c8153288b3c4f2ec64af8ff152f65","tests/integration/ui/compile_fail/attr_with.stderr":"7ba12a9ef37bc83db0ee8ebb14e31ba5618f8995128b32ebd917bbe729ed49b6","tests/integration/ui/compile_fail/broken_intra_doc_links.rs":"320cc747edc37cc70d2e42a289dd8917b65698bf400a45e0e95e000f238e12d2","tests/integration/ui/compile_fail/broken_intra_doc_links.stderr":"f16d97bbb7555890550057a025eeeeb03f9077d0201969eeaf95a89dfeae5aea","tests/integration/ui/compile_fail/collections.rs":"a3a62fb6a5437f9e6a70ee0506a025d69268f3e81af396027bf41ec5172288b7","tests/integration/ui/compile_fail/collections.stderr":"66a7bc8ed31ccb571c20c3d91018d13ffe36c95a67fddf738a0b8f285e39ca7c","tests/integration/ui/compile_fail/derive_builder.rs":"565fd21853ecb858f8dd3304e487e483d333367bd423a41eea9fbd91a4452b85","tests/integration/ui/compile_fail/derive_builder.stderr":"27e485795c9d13df50ddea386a6348f62ae8a9f3aa2298a30b0dc20fd56890d0","tests/integration/ui/compile_fail/diagnostic_on_unimplemented.rs":"b920f5964299bf8c5cf51b98756207e7f9e025aed59514964ad00770ebefc86a","tests/integration/ui/compile_fail/diagnostic_on_unimplemented.stderr":"6482f1037076e8e2e4e9ab0c671df91ddbb8efae1f1d90ac13489afa41be73f1","tests/integration/ui/compile_fail/members_order.rs":"a6efe9c32aac6cea9da3865828995d67898b4ba1a58ddb3797016fb6a979e835","tests/integration/ui/compile_fail/members_order.stderr":"45729aa18e82ede2ad7fdb3dea0604bcf89e6f5139725b2ad63435adf30e92ed","tests/integration/ui/compile_fail/name_conflicts.rs":"3bccea10dda185fe778d055f25bd0ead3fcff28fe341108fcf822a4e0295ad7a","tests/integration/ui/compile_fail/name_conflicts.stderr":"000b513ecd86e1b4f047090fed22b949fd9b781ef9a693dbdb01bb12e0ae1420","tests/integration/ui/compile_fail/positional_members.rs":"6a6f83c98d6b9be41c7a6393b767f48ab2bb250ab20ca5395e296c0ddc8a9550","tests/integration/ui/compile_fail/positional_members.stderr":"501bf7a1f6c71d2ae126609b0a6bff1599966ed2f6f2e2dfd4ba282292810d35","tests/integration/ui/compile_fail/private_fields_access.rs":"3ff0dfb64f9da292e0c25dad373c62f5ea67b524c049e076c8f554c3971b60bc","tests/integration/ui/compile_fail/private_fields_access.stderr":"9a5003dca41e2f967154b5e9efe034bde3ec48cecf5839e4c78b23d8e3b28a43","tests/integration/ui/compile_fail/warnings.rs":"53124ff50971ef4086f87b7ae72dfab0a3f3309bb7aa3024e035a70b3082a46c","tests/integration/ui/compile_fail/warnings.stderr":"35bd89029da20959e0424d6a6e60fffbe312ed4317bf766c6fc09be8ce364bb6","tests/integration/ui/compile_fail/wrong_delimiters.rs":"f27ef77bf68369150609f02f1a51a54cb882971d33fee4da091413c10c9f431b","tests/integration/ui/compile_fail/wrong_delimiters.stderr":"400704dd0587f96c78d3ed00138b060efe288fd124cd0c4bfe890e456b05b9f6","tests/integration/ui/mod.rs":"62ded0b6bf908bec2164de93d480843d9f7de6fd7a2d53ee964d2826a1679e55"},"package":"fe7acc34ff59877422326db7d6f2d845a582b16396b6b08194942bf34c6528ab"}