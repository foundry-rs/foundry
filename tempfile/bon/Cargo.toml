# THIS FILE IS AUTOMATICALLY GENERATED BY CARGO
#
# When uploading crates to the registry Cargo will automatically
# "normalize" Cargo.toml files for maximal compatibility
# with all versions of Cargo and also rewrite `path` dependencies
# to registry (e.g., crates.io) dependencies.
#
# If you are reading this file be aware that the original Cargo.toml
# will likely look very different (and much more reasonable).
# See Cargo.toml.orig for the original contents.

[package]
edition = "2021"
rust-version = "1.59.0"
name = "bon"
version = "3.3.2"
build = false
autolib = false
autobins = false
autoexamples = false
autotests = false
autobenches = false
description = "Next-gen compile-time-checked builder generator, named function's arguments, and more!"
homepage = "https://bon-rs.com"
readme = "README.md"
keywords = [
    "builder",
    "macro",
    "derive",
    "constructor",
    "setter",
]
categories = [
    "rust-patterns",
    "data-structures",
    "asynchronous",
    "no-std",
    "no-std::no-alloc",
]
license = "MIT OR Apache-2.0"
repository = "https://github.com/elastio/bon"

[package.metadata.docs.rs]
all-features = true
rustdoc-args = ["--generate-link-to-definition"]
targets = ["x86_64-unknown-linux-gnu"]

[lib]
name = "bon"
path = "src/lib.rs"

[[test]]
name = "integration"
path = "tests/integration/main.rs"

[dependencies.bon-macros]
version = "=3.3.2"

[dependencies.rustversion]
version = "1"

[dev-dependencies.expect-test]
version = "1.4.1"

[dev-dependencies.tokio]
version = "1.29.1"
features = [
    "macros",
    "rt-multi-thread",
]

[dev-dependencies.trybuild]
version = "1.0.89"

[features]
alloc = []
default = ["std"]
experimental-getter = ["bon-macros/experimental-getter"]
experimental-overwritable = ["bon-macros/experimental-overwritable"]
implied-bounds = ["bon-macros/implied-bounds"]
std = ["alloc"]

[lints.clippy]
assertions_on_result_states = "warn"
await_holding_refcell_ref = "warn"
bool_to_int_with_if = "warn"
borrow_as_ptr = "warn"
branches_sharing_code = "warn"
cast_lossless = "warn"
cast_ptr_alignment = "warn"
checked_conversions = "warn"
clear_with_drain = "warn"
cloned_instead_of_copied = "warn"
collection_is_never_read = "warn"
copy_iterator = "warn"
dbg_macro = "warn"
debug_assert_with_mut_call = "warn"
default_union_representation = "warn"
doc_link_with_quotes = "warn"
doc_markdown = "warn"
empty_drop = "warn"
empty_line_after_doc_comments = "warn"
empty_line_after_outer_attr = "warn"
expl_impl_clone_on_copy = "warn"
explicit_into_iter_loop = "warn"
explicit_iter_loop = "warn"
filter_map_next = "warn"
flat_map_option = "warn"
float_cmp = "warn"
fn_params_excessive_bools = "warn"
fn_to_numeric_cast_any = "warn"
format_push_string = "warn"
if_then_some_else_none = "warn"
implicit_clone = "warn"
implicit_saturating_sub = "warn"
imprecise_flops = "warn"
index_refutable_slice = "warn"
invalid_upcast_comparisons = "warn"
iter_not_returning_iterator = "warn"
iter_on_empty_collections = "warn"
iter_with_drain = "warn"
large_include_file = "warn"
linkedlist = "warn"
lossy_float_literal = "warn"
manual_clamp = "warn"
manual_instant_elapsed = "warn"
manual_let_else = "allow"
manual_ok_or = "warn"
many_single_char_names = "warn"
map_flatten = "warn"
map_unwrap_or = "allow"
match_on_vec_items = "warn"
match_wild_err_arm = "warn"
maybe_infinite_iter = "warn"
mixed_read_write_in_expression = "warn"
mut_mut = "warn"
naive_bytecount = "warn"
needless_bitwise_bool = "warn"
needless_collect = "warn"
needless_continue = "warn"
needless_for_each = "warn"
negative_feature_names = "warn"
no_effect_underscore_binding = "warn"
no_mangle_with_rust_abi = "warn"
non_send_fields_in_send_ty = "warn"
nonstandard_macro_braces = "warn"
or_fun_call = "warn"
path_buf_push_overwrite = "warn"
ptr_as_ptr = "warn"
range_minus_one = "warn"
range_plus_one = "warn"
rc_buffer = "warn"
rc_mutex = "warn"
read_zero_byte_vec = "warn"
redundant_clone = "warn"
redundant_feature_names = "warn"
ref_binding_to_reference = "warn"
ref_option_ref = "warn"
rest_pat_in_fully_bound_structs = "warn"
same_functions_in_if_condition = "warn"
string_lit_as_bytes = "warn"
string_to_string = "warn"
suboptimal_flops = "warn"
suspicious_operation_groupings = "warn"
suspicious_xor_used_as_pow = "warn"
trailing_empty_array = "warn"
trait_duplication_in_bounds = "warn"
transmute_ptr_to_ptr = "warn"
transmute_undefined_repr = "warn"
try_err = "warn"
type_repetition_in_bounds = "warn"
unchecked_duration_subtraction = "warn"
unicode_not_nfc = "warn"
unnecessary_join = "warn"
unnecessary_self_imports = "warn"
unnecessary_struct_initialization = "warn"
unnecessary_wraps = "warn"
unnested_or_patterns = "warn"
unused_async = "warn"
unused_peekable = "warn"
unused_rounding = "warn"
useless_let_if_seq = "warn"
useless_transmute = "warn"
verbose_bit_mask = "warn"
wildcard_dependencies = "warn"
zero_sized_map_values = "warn"

[lints.clippy.nursery]
level = "warn"
priority = -2

[lints.clippy.pedantic]
level = "warn"
priority = -1

[lints.rust]
deprecated_safe = "warn"
elided_lifetimes_in_paths = "warn"
explicit_outlives_requirements = "warn"
ffi_unwind_calls = "warn"
let_underscore_drop = "warn"
macro_use_extern_crate = "warn"
meta_variable_misuse = "warn"
missing_abi = "warn"
missing_debug_implementations = "warn"
missing_docs = "warn"
missing_unsafe_on_extern = "warn"
non_ascii_idents = "warn"
non_local_definitions = "warn"
redundant_lifetimes = "warn"
single_use_lifetimes = "warn"
trivial_casts = "warn"
trivial_numeric_casts = "warn"
unit_bindings = "warn"
unnameable_types = "warn"
unreachable_pub = "warn"
unsafe_code = "warn"
unstable_features = "warn"
unused_extern_crates = "warn"
unused_import_braces = "warn"
unused_lifetimes = "warn"
unused_macro_rules = "warn"
unused_qualifications = "warn"
variant_size_differences = "warn"

[lints.rust.rust_2024_compatibility]
level = "warn"
priority = 1

[lints.rust.tail_expr_drop_order]
level = "allow"
priority = 2

[lints.rust.unexpected_cfgs]
level = "warn"
priority = 0
check-cfg = ["cfg(nightly)"]

[lints.rustdoc]
missing_crate_level_docs = "warn"
unescaped_backticks = "warn"
