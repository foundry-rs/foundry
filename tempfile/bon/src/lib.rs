#![doc(
    html_logo_url = "https://bon-rs.com/bon-logo-thumb.png",
    html_favicon_url = "https://bon-rs.com/bon-logo-medium.png"
)]
#![doc = include_str!("../README.md")]
#![cfg_attr(not(feature = "std"), no_std)]
#![cfg_attr(docsrs, feature(doc_cfg))]
// We mark all items from the `private` module as deprecated to signal that they are
// implementation details and should not be used directly. Unfortunately, this triggers
// the deprecation warnings within this crate itself everywhere we use them, so we just
// suppress this lint for the entire crate.
#![allow(deprecated)]

// Rexport all macros from the proc-macro crate.
pub use bon_macros::{bon, builder, map, set, Builder};

/// Small utility declarative macros for creating colletions with [`Into`] conversions.
mod collections;

#[doc(hidden)]
#[deprecated = "the items from the `bon::__` module are an implementation detail; \
    they should not be used directly; if you found a need for this, then you are probably \
    doing something wrong; feel free to open an issue/discussion in our GitHub repository \
    (https://github.com/elastio/bon) or ask for help in our Discord server \
    (https://bon-rs.com/discord)"]
pub mod __;

mod builder_state;
