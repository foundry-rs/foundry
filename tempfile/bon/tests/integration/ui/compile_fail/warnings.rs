#![deny(warnings)]
use bon::{bon, builder, Builder};

fn main() {
    // Test #[must_use] for setters
    {
        #[allow(dead_code)]
        #[derive(Builder)]
        struct Example {
            x: u32,
            y: u32,
        }

        #[bon]
        impl Example {
            #[builder]
            #[must_use]
            fn must_use() -> u32 {
                99
            }
        }

        #[builder]
        #[must_use]
        fn must_use() -> u32 {
            99
        }

        Example::builder();
        Example::must_use();
        must_use();

        Example::builder().x(1);
        Example::builder().x(1).y(2).build();

        Example::must_use().call();

        must_use().call();
        __orig_must_use();

        #[builder]
        #[cfg_attr(all(), must_use = "must use message")]
        fn must_use_under_cfg() -> u32 {
            99
        }

        must_use_under_cfg().call();

        #[builder]
        #[cfg_attr(any(), must_use = "unreachable must use")]
        fn must_use_compiled_out() -> u32 {
            99
        }

        must_use_compiled_out().call();
    }

    // Test #[must_use] for getters
    {
        #[derive(Builder)]
        struct Sut {
            #[builder(getter)]
            x1: u32,

            #[builder(getter, default)]
            x2: u32,

            #[builder(getter)]
            x3: Option<u32>,
        }

        let builder = Sut::builder().x1(1).x2(2).x3(3);

        // Make sure there are `#[must_use]` warnings
        builder.get_x1();
        builder.get_x2();
        builder.get_x3();
    }
}
