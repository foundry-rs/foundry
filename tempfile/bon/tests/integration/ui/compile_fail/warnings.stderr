error: unused `ExampleBuilder` that must be used
  --> tests/integration/ui/compile_fail/warnings.rs:29:9
   |
29 |         Example::builder();
   |         ^^^^^^^^^^^^^^^^^^
   |
   = note: the builder does nothing until you call `build()` on it to finish building
note: the lint level is defined here
  --> tests/integration/ui/compile_fail/warnings.rs:1:9
   |
1  | #![deny(warnings)]
   |         ^^^^^^^^
   = note: `#[deny(unused_must_use)]` implied by `#[deny(warnings)]`
help: use `let _ = ...` to ignore the resulting value
   |
29 |         let _ = Example::builder();
   |         +++++++

error: unused `ExampleMustUseBuilder` that must be used
  --> tests/integration/ui/compile_fail/warnings.rs:30:9
   |
30 |         Example::must_use();
   |         ^^^^^^^^^^^^^^^^^^^
   |
   = note: the builder does nothing until you call `call()` on it to finish building
help: use `let _ = ...` to ignore the resulting value
   |
30 |         let _ = Example::must_use();
   |         +++++++

error: unused `MustUseBuilder` that must be used
  --> tests/integration/ui/compile_fail/warnings.rs:31:9
   |
31 |         must_use();
   |         ^^^^^^^^^^
   |
   = note: the builder does nothing until you call `call()` on it to finish building
help: use `let _ = ...` to ignore the resulting value
   |
31 |         let _ = must_use();
   |         +++++++

error: unused `ExampleBuilder` that must be used
  --> tests/integration/ui/compile_fail/warnings.rs:33:9
   |
33 |         Example::builder().x(1);
   |         ^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: the builder does nothing until you call `build()` on it to finish building
help: use `let _ = ...` to ignore the resulting value
   |
33 |         let _ = Example::builder().x(1);
   |         +++++++

error: unused return value of `ExampleBuilder::<S>::build` that must be used
  --> tests/integration/ui/compile_fail/warnings.rs:34:9
   |
34 |         Example::builder().x(1).y(2).build();
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: building a struct without using it is likely a bug
help: use `let _ = ...` to ignore the resulting value
   |
34 |         let _ = Example::builder().x(1).y(2).build();
   |         +++++++

error: unused return value of `ExampleMustUseBuilder::<S>::call` that must be used
  --> tests/integration/ui/compile_fail/warnings.rs:36:9
   |
36 |         Example::must_use().call();
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
help: use `let _ = ...` to ignore the resulting value
   |
36 |         let _ = Example::must_use().call();
   |         +++++++

error: unused return value of `MustUseBuilder::<S>::call` that must be used
  --> tests/integration/ui/compile_fail/warnings.rs:38:9
   |
38 |         must_use().call();
   |         ^^^^^^^^^^^^^^^^^
   |
help: use `let _ = ...` to ignore the resulting value
   |
38 |         let _ = must_use().call();
   |         +++++++

error: unused return value of `__orig_must_use` that must be used
  --> tests/integration/ui/compile_fail/warnings.rs:39:9
   |
39 |         __orig_must_use();
   |         ^^^^^^^^^^^^^^^^^
   |
help: use `let _ = ...` to ignore the resulting value
   |
39 |         let _ = __orig_must_use();
   |         +++++++

error: unused return value of `MustUseUnderCfgBuilder::<S>::call` that must be used
  --> tests/integration/ui/compile_fail/warnings.rs:47:9
   |
47 |         must_use_under_cfg().call();
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: must use message
help: use `let _ = ...` to ignore the resulting value
   |
47 |         let _ = must_use_under_cfg().call();
   |         +++++++

error: unused return value of `SutBuilder::<S>::get_x1` that must be used
  --> tests/integration/ui/compile_fail/warnings.rs:75:9
   |
75 |         builder.get_x1();
   |         ^^^^^^^^^^^^^^^^
   |
   = note: this method has no side effects; it only returns a value
help: use `let _ = ...` to ignore the resulting value
   |
75 |         let _ = builder.get_x1();
   |         +++++++

error: unused return value of `SutBuilder::<S>::get_x2` that must be used
  --> tests/integration/ui/compile_fail/warnings.rs:76:9
   |
76 |         builder.get_x2();
   |         ^^^^^^^^^^^^^^^^
   |
   = note: this method has no side effects; it only returns a value
help: use `let _ = ...` to ignore the resulting value
   |
76 |         let _ = builder.get_x2();
   |         +++++++

error: unused return value of `SutBuilder::<S>::get_x3` that must be used
  --> tests/integration/ui/compile_fail/warnings.rs:77:9
   |
77 |         builder.get_x3();
   |         ^^^^^^^^^^^^^^^^
   |
   = note: this method has no side effects; it only returns a value
help: use `let _ = ...` to ignore the resulting value
   |
77 |         let _ = builder.get_x3();
   |         +++++++
